File: packages/nlu/kotlin/nlukit/src/main/kotlin/com/bridgething/nlukit/NluSlotMapping.kt
package com.bridgething.nlukit

import com.bridgething.schema.NluAmount
import com.bridgething.schema.NluDirection
import com.bridgething.schema.NluPhoneAction
import com.bridgething.schema.NluPlaybackSpeed
import com.bridgething.schema.NluPopularityFilter
import com.bridgething.schema.NluRepeatMode
import com.bridgething.schema.NluScope
import com.bridgething.schema.NluSlots
import com.bridgething.schema.NluTargetType
import com.bridgething.schema.NluView
import uniffi.nlu.SlotValue

object NluSlotMapping {
    fun apply(slots: List<SlotValue>): NluSlots =
        slots.fold(NluSlots()) { out, slot ->
            val v = slot.value
            when (slot.name) {
                "target" -> out.copy(target = v)
                "playlist" -> out.copy(playlist = v)
                "genre" -> out.copy(genre = v)
                "mood" -> out.copy(mood = v)
                "era" -> out.copy(era = v)
                "webapp_name" -> out.copy(webappName = v)
                "preset" -> out.copy(preset = v)
                "target_type" -> out.copy(targetType = enum(NluTargetType.entries, camel(v)) { it.string })
                "popularity_filter" -> out.copy(popularityFilter = enum(NluPopularityFilter.entries, camel(v)) { it.string })
                "scope" -> out.copy(scope = enum(NluScope.entries, camel(v)) { it.string })
                "view" -> out.copy(view = enum(NluView.entries, camel(v)) { it.string })
                "repeat_mode" -> out.copy(repeatMode = enum(NluRepeatMode.entries, camel(v)) { it.string })
                "speed" -> out.copy(speed = enum(NluPlaybackSpeed.entries, v) { it.string })
                "direction" -> out.copy(direction = enum(NluDirection.entries, camel(v)) { it.string })
                "amount" -> out.copy(amount = enum(NluAmount.entries, camel(v)) { it.string })
                "phone_action" -> out.copy(phoneAction = enum(NluPhoneAction.entries, camel(v)) { it.string })
                "enabled" -> out.copy(enabled = bool(v))
                "mute" -> out.copy(mute = bool(v))
                "count" -> out.copy(count = v.toUIntOrNull())
                "position" -> out.copy(position = v.toUIntOrNull())
                "level" -> out.copy(level = v.toUIntOrNull())
                "seconds" -> out.copy(seconds = v.toIntOrNull())
                else -> out
            }
        }

    fun camel(token: String): String {
        val parts = token.split("_")
        if (parts.isEmpty()) return token
        return parts.first() + parts.drop(1).joinToString("") { it.replaceFirstChar(Char::uppercaseChar) }
    }

    fun bool(token: String): Boolean? = when (token.lowercase()) {
        "true" -> true
        "false" -> false
        else -> null
    }

    private fun <T> enum(entries: List<T>, token: String, wire: (T) -> String): T? =
        entries.firstOrNull { wire(it) == token }
}

File: packages/nlu/swift/Sources/BridgethingNluKit/NluSlotMapping.swift
import BridgethingCompanion
import BridgethingSchema
import Foundation
import Nlu

public enum NluSlotMapping {
    public static func apply(_ slots: [SlotValue]) -> NluMutableSlots {
        var out = NluMutableSlots()
        for slot in slots {
            let v = slot.value
            switch slot.name {
            case "target": out.target = v
            case "playlist": out.playlist = v
            case "genre": out.genre = v
            case "mood": out.mood = v
            case "era": out.era = v
            case "webapp_name": out.webappName = v
            case "preset": out.preset = v
            case "target_type": out.targetType = NluTargetType(rawValue: camel(v))
            case "popularity_filter": out.popularityFilter = NluPopularityFilter(rawValue: camel(v))
            case "scope": out.scope = NluScope(rawValue: camel(v))
            case "view": out.view = NluView(rawValue: camel(v))
            case "repeat_mode": out.repeatMode = NluRepeatMode(rawValue: camel(v))
            case "speed": out.speed = NluPlaybackSpeed(rawValue: v)
            case "direction": out.direction = NluDirection(rawValue: camel(v))
            case "amount": out.amount = NluAmount(rawValue: camel(v))
            case "phone_action": out.phoneAction = NluPhoneAction(rawValue: camel(v))
            case "enabled": out.enabled = bool(v)
            case "mute": out.mute = bool(v)
            case "count": out.count = UInt32(v)
            case "position": out.position = UInt32(v)
            case "level": out.level = UInt32(v)
            case "seconds": out.seconds = Int32(v)
            default: break
            }
        }
        return out
    }

    static func camel(_ token: String) -> String {
        let parts = token.split(separator: "_")
        guard let first = parts.first else { return token }
        return String(first) + parts.dropFirst().map { $0.prefix(1).uppercased() + $0.dropFirst() }.joined()
    }

    static func bool(_ token: String) -> Bool? {
        switch token.lowercased() {
        case "true": return true
        case "false": return false
        default: return nil
        }
    }
}
